### ISSUE TITLE
ESbuild slower incremental rebuilds

### ISSUE THREAD

**Issue by vbraun** [angular/angular-cli#25935]
### Command

serve

### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

Incremental rebuilds with the new `@angular-devkit/build-angular:browser-esbuild` are a bit slower for us than the old webpack dev server, especially for minor changes at the leaves of the import graph. This is unexpected as esbuild's sole raison d'être is to be fast ;)

It is understood that the first incremental rebuild is slower than subsequent rebuilds, so for all tests we tigger the incremental build multiple times until times stabilize.

Our app has >1000 components in >100 libraries. For testing, I just `touch` a typescript file in a feature library. So very few files should have to be recompiled & rebundled. With esbuild I get 11 sec for incremental builds, with webpack 9 seconds. 

Looking at running processes I see esbuild pegging most CPU cores at 100% for most of the time during the incremental rebuild. This seems to be corroborated by enabling `NG_BUILD_DEBUG_PERF`, which only logs negligible times:
```
$ NG_BUILD_DEBUG_PERF=1 ng serve
[...]
⠋ Changes detected. Rebuilding...One or more browsers which are configured in the project's Browserslist configuration will be ignored as ES5 output is not supported by the Angular CLI.
Ignored browsers: kaios 2.5, op_mini all
DURATION[NG_READ_CONFIG]: 0.018551948s
DURATION[NG_CREATE_PROGRAM]: 0.237099920s
⠙ Changes detected. Rebuilding...
  mobile-list-view.component.css  284b 

⚡ Done in 0ms
DURATION[NG_ANALYZE_PROGRAM]: 0.616578717s
DURATION[NG_FIND_AFFECTED]: 0.000006410s
DURATION[NG_DIAGNOSTICS_TOTAL]: 0.021461028s
DURATION[NG_EMIT_TS]: 0.064279846s
DURATION[NG_DIAGNOSTICS_SYNTACTIC]: 0.001562300s [count: 11431; avg: 0.000000137s; min: 0.000000070s; max: 0.000005060s]
DURATION[NG_DIAGNOSTICS_SEMANTIC]: 0.004640593s [count: 11431; avg: 0.000000406s; min: 0.000000100s; max: 0.000057390s]
DURATION[NG_EMIT_JS*]: 0.002667022s [count: 34; avg: 0.000078442s; min: 0.000003780s; max: 0.000820767s]
⠸ Changes detected. Rebuilding...
[... snip chunks output...]
Application bundle generation complete. [11.095 seconds]
```

What is esbuild doing? Unfortunately there is no official profiling support, but it seems there is a lot of filesystem churn. When attaching strace to the running esbuild server (pid 698557 here), I get a surprising number of stat calls. For starters:
```
strace -f -p 698557 |& grep /home/vbraun/frontend/package.json
[pid 698634] newfstatat(AT_FDCWD, "/home/vbraun/frontend/package.json",  <unfinished ...>
[...]
```
gives 71083 `stat()` calls to the `package.json` file from the different PIDs of esbuild's thread pool, each time an incremental build is triggered. Stat should run in a few microseconds but that alone probably adds up to more than 0.1s. 

This is how far I have gotten debugging this..

### Minimal Reproduction

This is a private repo but others have reported the same: https://github.com/angular/angular-cli/issues/25130#issuecomment-1684157200

### Your Environment

```text
$ npx ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 16.2.3
Node: 16.18.1
Package Manager: npm 8.19.2
OS: linux x64

Angular: 16.2.6
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, localize, platform-browser
... platform-browser-dynamic, router, upgrade

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1602.3
@angular-devkit/build-angular   16.2.3
@angular-devkit/core            16.2.3
@angular-devkit/schematics      16.2.3
@angular/cdk                    16.2.5
@angular/cli                    16.2.3
@angular/material               16.2.5
@schematics/angular             15.2.6
ng-packagr                      16.2.3
rxjs                            7.8.1
typescript                      4.9.5
zone.js                         0.13.3
```

### Anything else relevant?

* Fedora 38 x86_64, btrfs
* Ryzen 7900X 12-Core Processor, 64 GB RAM, nvme SSD
* Using `NODE_OPTIONS=--max_old_space_size=8192` to avoid memory bottlenecks

**Comment by csvn** [comment]
I had the same issue ~4months ago when I tried the esbuild-builder. For us esbuild was consistently about 1s slower than Webpack, changing a file with some whitespace changes. I just left that PR with the thought that this might be improved on the road to non-developer preview.

**Comment by LanderBeeuwsaert** [comment]
Jep, same here, we had the same

**Comment by alan-agius4** [comment]
There has been several performance improvements in version 17.0.0-next.8, can you please check if the problem persists?

**Comment by vbraun** [comment]
I tried 17.0.0-next.8 and it is indeed much faster :rocket: 
```
Application bundle generation complete. [3.807 seconds]
```
Moreover, it now only checks (stat syscall) 32 times that `package.json` is still there, and not 70k times :tada:

**Comment by alan-agius4** [comment]
Thanks for the confirmation. Closing se the issue appears to have been solved.

**Comment by csvn** [comment]
It's a lot better for me too! Instead of being ~1s slower, in watch mode it's now ~200ms faster with esbuild on incremental builds (if first change is discounted). `ng build` is still ~2s slower with esbuild (~22s vs 24s, with build-cache). But watch mode is more important, and this issue was regarding slower incremental builds.

TL;DR: I agree that we can consider this issue closed 😇
